### PR TITLE
Add ClickHouse

### PR DESCRIPTION
**Pull request by pjhampton**
This PR introduces ClickHouse support by finishing off https://github.com/googleapis/genai-toolbox/pull/713. We have had quite a few requests internally and externally of people who want ClickHouse support and wish to contribute but can't use the toolbox because of this stalled PR. We also tried to reach out to @sidpan1 ourselves with no success so we are going to assume the contributor doesn't have time to complete. We would like to thank @sidpan1 for getting this going - community contributions help make ClickHouse better for everyone. 

This PR introduces

  - Adds ClickHouse HTTPS and HTTP source support (drops native support and compression in connection)
  - Adds tools: `describe_table`, `execute_sql`, `list_databases`, `list_tables`
  - Adds unit and integration tests
  - Adds docs

I didn't understand you CI. If you can do it yourselves that would be great, otherwise give more detailed intructions than what is in the CONTRIBUTING.md.
Other contributors can build on this foundation

### Connect to local ClickHouse

```sh
CLICKHOUSE_HOST="localhost" \
  CLICKHOUSE_PORT=8123 \
  CLICKHOUSE_USER="default" \
  CLICKHOUSE_PASSWORD="" \
  CLICKHOUSE_DATABASE=default \
  CLICKHOUSE_PROTOCOL="http" \
  go run main.go --prebuilt clickhouse --ui --port 8080
```

### Connect to ClickHouse Cloud

```sh
CLICKHOUSE_HOST="tsmtweovmw.us-east-2.aws.clickhouse.cloud" \
  CLICKHOUSE_PORT=8443 \
  CLICKHOUSE_USER="default" \
  CLICKHOUSE_PASSWORD="[REDACTED]" \
  CLICKHOUSE_DATABASE=default \
  CLICKHOUSE_PROTOCOL=https \
  go run main.go --prebuilt clickhouse --ui --port 8080
```
Screenshot 2025-08-06 at 10 01 01

### CODE REVIEW

**Comment by duwenxin99** [comment]
Hi @pjhampton, thanks for opening the PR to continue this work!
1. Could you separate the ClickHouse Source and each tool (with their corresponding docs & tests) into separate smaller PRs? This would be extremely helpful for our reviews and make sure we are not missing on details. It also looks better on the release note.
2. For the integration test, you will need to call the following generic tests with the `clickhouse-sql` tool:
- [RunToolGetTest](https://github.com/googleapis/genai-toolbox/blob/fd300dc606d88bf9f7bba689e2cee4e3565537dd/tests/tool.go#L31): tests for the GET endpoint that returns the tool's manifest.
- [RunToolInvokeTest](https://github.com/googleapis/genai-toolbox/blob/fd300dc606d88bf9f7bba689e2cee4e3565537dd/tests/tool.go#L79): tests for tool calling through the native Toolbox endpoints.
- [RunMCPToolCallMethod](https://github.com/googleapis/genai-toolbox/blob/fd300dc606d88bf9f7bba689e2cee4e3565537dd/tests/tool.go#L554): tests tool calling through the MCP endpoints.

Run this for the `clickhouse-execute-sql` tool:
- [RunExecuteSqlToolInvokeTest](https://github.com/googleapis/genai-toolbox/blob/fd300dc606d88bf9f7bba689e2cee4e3565537dd/tests/tool.go#L431)

The way to run the generic test suite is to [set up a tool config file](https://github.com/googleapis/genai-toolbox/blob/fd300dc606d88bf9f7bba689e2cee4e3565537dd/tests/common.go#L31) and input expected outputs for the tool requests to the[ test function](https://github.com/googleapis/genai-toolbox/blob/fd300dc606d88bf9f7bba689e2cee4e3565537dd/tests/tool.go#L79). This function will invoke different tools set up in the previous step with different parameters, and check the result with the expected output. You can look at its test table for specific request content and determine what the output should be for ClickHouse. Feel free to reuse these [helper functions ](https://github.com/googleapis/genai-toolbox/blob/fd300dc606d88bf9f7bba689e2cee4e3565537dd/tests/common.go#L338) if they are applicable to ClickHouse. You can use the [Postgres integration test](https://github.com/googleapis/genai-toolbox/blob/fd300dc606d88bf9f7bba689e2cee4e3565537dd/tests/postgres/postgres_integration_test.go) as an example.

For the rest of the tools, you should write your own integration tests and make sure to cover the edge cases.
To run the integration test on Github, make sure to move the `.ci/integration.cloudbuild.yaml` file change over from https://github.com/googleapis/genai-toolbox/pull/713.
Let me know of any questions. Thanks!

**Comment by pjhampton** [comment]
> Hi @pjhampton, thanks for opening the PR to continue this work!

No worries [duwenxin99](https://github.com/duwenxin99) - happy to help. 

> Could you separate the ClickHouse Source and each tool (with their corresponding docs & tests) into separate smaller PRs? This would be extremely helpful for our reviews and make sure we are not missing on details. It also looks better on the release note.

Got it - while not unreasonable, I am hoping you will be amenable to keeping it how it is in the interest of time. This is a net new feature, and the code is incredibly basic along with the tools. I'm not sure how practical it is as I don't have the time to engage in multiple review cycles and it seems this didn't apply to all other integrations, looking through the history.  Let's continue unless absolutely nessesary.

> For the integration test, you will need to call the following generic tests with the clickhouse-sql tool:

You got it.

**Comment by duwenxin99** [comment]
Hi @pjhampton, thanks for the test update! You are correct that we didn't enforce this rule 100% previously and that was why we had to do fixes afterwards multiple times (because it was easy to overlook the details in a PR with thousands of lines), which turned out to be more time-consuming for us and the contributors. However, I do understand that it's hard to keep track of so many PRs at the same time since we are trying to add 5 new tools, so we could instead separate them into 2 PRs:
1. For the first PR (we can just use the current one), add the initial ClickHouse Source and basic tools (clickhouse-sql and clickhouse-execute-sql).
2. Add the rest of the tools and their docs in the second PR.
Btw could you also mention me when you need reviews so that I can notice it faster? This is because I'm currently keeping track of over 5 integrations at the same time. I don't want to miss it and keep you waiting. Thank you!

**Comment by pjhampton** [comment]
@duwenxin99 for sure - no worries. I can follow up with the tools in another PR. Thanks for understanding and look forward to your comments.

**Comment by pjhampton** [comment]
Hey @duwenxin99 - just checking to see if there is anything I can do to help get this over the line and raise the tools PRs?

File: internal/sources/clickhouse/clickhouse.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+type Config struct {
+	Name     string `yaml:"name" validate:"required"`
+	Kind     string `yaml:"kind" validate:"required"`
+	Host     string `yaml:"host" validate:"required"`
+	Port     string `yaml:"port" validate:"required"`
+	User     string `yaml:"user" validate:"required"`
+	Password string `yaml:"password"`
+	Database string `yaml:"database" validate:"required"`
```
Comment (duwenxin99) [inline internal/sources/clickhouse/clickhouse.go]:
```suggestion
    Database string `yaml:"database" validate:"required"`
	User     string `yaml:"user" validate:"required"`
	Password string `yaml:"password"`
```

File: internal/sources/clickhouse/clickhouse_test.go [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+func TestNewConfig(t *testing.T) {
+	tests := []struct {
+		name     string
+		yaml     string
+		expected Config
+	}{
+		{
+			name: "all fields specified",
+			yaml: `
+name: test-clickhouse
```
Comment (duwenxin99) [inline internal/sources/clickhouse/clickhouse_test.go]: Could you correct the fields & indentation for the yaml inputs? [Example](https://github.com/googleapis/genai-toolbox/blob/e92fed63decd5af6d7319b120d10dfccfcc33f66/internal/sources/postgres/postgres_test.go#L37)
It should be something like:
```
sources:
    my-clickhouse-instance:
        kind: clickhouse
        host: 127.0.0.1
        port: "8123"
        user: testuser
        database: testdb
```

File: internal/tools/clickhouse/clickhouseexecutesql.go [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+	Parameters   tools.Parameters `yaml:"parameters"`
+
+	Pool        *sql.DB
+	manifest    tools.Manifest
+	mcpManifest tools.McpManifest
+}
+
+func (t ExecuteSQLTool) Invoke(ctx context.Context, params tools.ParamValues) (any, error) {
+	sliceParams := params.AsSlice()
+	sqlStatement, ok := sliceParams[0].(string)
```
Comment (duwenxin99) [inline internal/tools/clickhouse/clickhouseexecutesql.go]: Use maps to get values instead of slices ([Example](https://github.com/googleapis/genai-toolbox/blob/e92fed63decd5af6d7319b120d10dfccfcc33f66/internal/tools/bigquery/bigqueryexecutesql/bigqueryexecutesql.go#L130))

File: internal/tools/clickhouse/clickhousesql.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+		Pool:               s.ClickHousePool(),
+		manifest:           tools.Manifest{Description: cfg.Description, Parameters: paramManifest, AuthRequired: cfg.AuthRequired},
+		mcpManifest:        mcpManifest,
+	}
+	return t, nil
+}
+
+var _ tools.Tool = SQLTool{}
+
+type SQLTool struct {
```
Comment (duwenxin99) [inline internal/tools/clickhouse/clickhousesql.go]: We should just name this struct `Tool` instead of `SQLTool`. Each tool should be in its separate package.

File: internal/tools/clickhouse/clickhousesql.go [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+	var out []any
+	for results.Next() {
+		err := results.Scan(values...)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse row: %w", err)
+		}
+		vMap := make(map[string]any)
+		for i, name := range cols {
+			switch colTypes[i].DatabaseTypeName() {
+			case "String", "FixedString":
```
Comment (duwenxin99) [inline internal/tools/clickhouse/clickhousesql.go]: Just want to confirm: do we need to convert any other types?

File: internal/tools/clickhouse/clickhousesql_test.go [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+}
+
+func TestNewSQLConfig(t *testing.T) {
+	ctx, err := testutils.ContextWithNewLogger()
+	if err != nil {
+		t.Fatalf("Failed to create context with logger: %v", err)
+	}
+
+	yamlContent := `
+name: test-clickhouse-tool
```
Comment (duwenxin99) [inline internal/tools/clickhouse/clickhousesql_test.go]: formatting is incorrect. Same as the previous comment.

File: internal/tools/clickhouse/clickhousesql_test.go [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+	manifest := tool.McpManifest()
+	if manifest.Name != "test-tool" {
+		t.Errorf("Expected name 'test-tool', got %s", manifest.Name)
+	}
+	if manifest.Description != "Test description" {
+		t.Errorf("Expected description 'Test description', got %s", manifest.Description)
+	}
+}
+
+func TestSQLTool_Authorized(t *testing.T) {
```
Comment (duwenxin99) [inline internal/tools/clickhouse/clickhousesql_test.go]: We should stick to camel cases

File: internal/tools/clickhouse/clickhousesql_test.go [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+	if err != nil {
+		t.Fatalf("Failed to create config: %v", err)
+	}
+
+	clickhouseConfig, ok := config.(SQLConfig)
+	if !ok {
+		t.Fatalf("Expected Config type, got %T", config)
+	}
+
+	if clickhouseConfig.Name != "test-clickhouse-tool" {
```
Comment (duwenxin99) [inline internal/tools/clickhouse/clickhousesql_test.go]: We should use a `want` value here to avoid all individual field checks ([example](https://github.com/googleapis/genai-toolbox/blob/e92fed63decd5af6d7319b120d10dfccfcc33f66/internal/tools/bigquery/bigquerysql/bigquery_test.go#L53))

File: internal/tools/clickhouse/common.go [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clickhouse
+
+import "database/sql"
+
+type compatibleSource interface {
```
Comment (duwenxin99) [inline internal/tools/clickhouse/common.go]: We can just move it to individual files if this is the only common code

File: tests/clickhouse/clickhouse_tools_integration_test.go [hunk 10]
```diff
@@ -0,0 +1,655 @@
+// Copyright 2025 Google LLC
```
Comment (duwenxin99) [inline tests/clickhouse/clickhouse_tools_integration_test.go]: Our integration test files should cover source + tool tests, so we don't need to separate them. Let's just move the content here to `clickhouse_integration_test.go`

**Review by duwenxin99 — CHANGES_REQUESTED** [review]
Hi @pjhampton, thanks for the patience! I left some comments. Feel free to let me know of any questions.

File: .ci/integration.cloudbuild.yaml [hunk 11]
```diff
@@ -594,6 +616,11 @@ availableSecrets:
       env: TIDB_USER
     - versionName: projects/$PROJECT_ID/secrets/tidb_pass/versions/latest
       env: TIDB_PASS
+      env: CLICKHOUSE_HOST
```
Comment (duwenxin99) [inline .ci/integration.cloudbuild.yaml:686]:
```suggestion
    - versionName: projects/$PROJECT_ID/secrets/clickhouse_hose/versions/latest
      env: CLICKHOUSE_HOST
```

File: internal/tools/clickhouse/clickhousesql.go (hunk 5 above)
Comment (pjhampton) [reply internal/tools/clickhouse/clickhousesql.go]: I believe this should be okay for now

**Comment by pjhampton** [comment]
Hey @duwenxin99 - keen for another review. Would be great to get this in and unblock folks wishing to use it.

**Comment by xyb** [comment]
@pjhampton Compilation failed with the latest code:
```
internal/tools/clickhouse/clickhousesql/clickhousesql.go:77:52: assignment mismatch: 3 variables but tools.ProcessParameters returns 4 values
```

Commit 8c0c603 works like a charm.

**Comment by pjhampton** [comment]
Thanks @xyb. These regressions have happened once or twice now because of upstream interface changes. 
@duwenxin99 can we merge this to prevent this from happening again? This has been open for 3 weeks now.

File: internal/tools/clickhouse/clickhouseexecutesql/clickhouseexecutesql.go [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+
+func newExecuteSQLConfig(ctx context.Context, name string, decoder *yaml.Decoder) (tools.ToolConfig, error) {
+	actual := ExecuteSQLConfig{Name: name}
+	if err := decoder.DecodeContext(ctx, &actual); err != nil {
+		return nil, err
+	}
+	return actual, nil
+}
+
+type ExecuteSQLConfig struct {
```
Comment (duwenxin99) [inline internal/tools/clickhouse/clickhouseexecutesql/clickhouseexecutesql.go]: This struct should just be Config since each tool is in a separate package

File: internal/tools/clickhouse/clickhousesql/clickhousesql.go [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+
+func newSQLConfig(ctx context.Context, name string, decoder *yaml.Decoder) (tools.ToolConfig, error) {
+	actual := SQLConfig{Name: name}
+	if err := decoder.DecodeContext(ctx, &actual); err != nil {
+		return nil, err
+	}
+	return actual, nil
+}
+
+type SQLConfig struct {
```
Comment (duwenxin99) [inline internal/tools/clickhouse/clickhousesql/clickhousesql.go]:
```suggestion
type Config struct {
```

File: tests/common.go [hunk 14]
```diff
@@ -510,6 +510,63 @@ func SetupMySQLTable(t *testing.T, ctx context.Context, pool *sql.DB, createStat
...
+// GetClickHouseSQLTmplToolStatement returns statements and param for template parameter test cases for clickhouse-sql kind
+func GetClickHouseSQLTmplToolStatement() (string, string) {
+	tmplSelectCombined := "SELECT * FROM {{.tableName}} WHERE id = ?"
+	tmplSelectFilterCombined := "SELECT * FROM {{.tableName}} WHERE {{.columnFilter}} = ?"
+	return tmplSelectCombined, tmplSelectFilterCombined
+}
+
+// SetupClickHouseSQLTable creates and inserts data into a table of tool
+// compatible with clickhouse-sql tool
+func SetupClickHouseSQLTable(t *testing.T, ctx context.Context, pool *sql.DB, createStatement, insertStatement, tableName string, params []any) func(*testing.T) {
```
Comment (duwenxin99) [inline tests/common.go]: We can just move all the clickhouse related functions to clickhouse_integration_test.go since they are not used by other sources.

File: internal/tools/clickhouse/clickhousesql/clickhousesql_test.go [hunk 15]
```diff
@@ -0,0 +1,283 @@
...
+	tests := []struct {
+		name    string
+		yaml    string
+		want    SQLConfig
+		wantErr bool
+	}{
+		{
+			name: "valid config with parameters",
+			yaml: `
+name: test-clickhouse-tool
```
Comment (duwenxin99) [inline internal/tools/clickhouse/clickhousesql/clickhousesql_test.go]: This and the test case below need reformatting. Should match `clickhouseexecutesql_test.go`

**Comment by duwenxin99** [comment]
Hi @pjhampton, sorry for the delay. The PR looks good to me other than a few small changes. After the test errors are fixed, we can go ahead and merge it. Thanks!

File: internal/prebuiltconfigs/prebuiltconfigs_test.go [hunk 16]
```diff
@@ -85,6 +86,7 @@ func TestGetPrebuiltTool(t *testing.T) {
 	alloydb_admin_config, _ := Get("alloydb-postgres-admin")
 	alloydb_config, _ := Get("alloydb-postgres")
 	bigquery_config, _ := Get("bigquery")
+	clickhouse_config, _ := Get("clickhouse")
```
Comment (duwenxin99) [inline internal/prebuiltconfigs/prebuiltconfigs_test.go:89]: Please also add to the root_test.go. Refer to the example [here](https://github.com/googleapis/genai-toolbox/pull/919/files#diff-558a3da4e777c67907421d3d7da4d33b537fa8fba7c600b2938bb2d39fdb59fa)

**Review by duwenxin99 — APPROVED** [review]

**Comment by pjhampton** [comment]
Thanks @duwenxin99!
